src/v/storage/segment_utils.cc

```diff
@@ -166,18 +166,42 @@ ss::future<segment_appender_ptr> make_segment_appender(
   ss::io_priority_class iopc,
   storage_resources& resources,
   std::optional<ntp_sanitizer_config> ntp_sanitizer_config) {
-    return internal::make_writer_handle(path, std::move(ntp_sanitizer_config))
-      .then([number_of_chunks, iopc, path, segment_size, &resources](
-              ss::file writer) {
+    return internal::make_writer_handle(path, std::nullopt)
+      .then([number_of_chunks,
+             iopc,
+             path,
+             segment_size,
+             &resources,
+             ntp_sanitizer_config = std::move(ntp_sanitizer_config)](
+              ss::file writer) mutable {
+          // file_io_sanitizer requires a pointer to the appender,
+          // so we create it inline and give it the pointer after
+          // the appender is created.
+          ss::shared_ptr<file_io_sanitizer> sanitized_writer{nullptr};
+          if (ntp_sanitizer_config) {
```

Comment:
> IIUC this passes the same config to every segment appender, which I believe correspond 1:1 with each segment.

Correct.

> Does that mean no matter how many segments are written, we will use the same seed on each segment within a given NTP?

Very good point! You're right. It's a tad clunky as each segment of an NTP will get the same sequence of failures.

> perhaps we should maintain some log/partition-wide seed generator that gets shared across all segments

I'd really like to avoid keeping any global state here. I'm thinking of hashing the ntp-seed with the filename, which would give us a unique seed for each segment appender.

Reply:
I ended up hashing the global seed with the file path. This should be deterministic, provided segment size jitter is disabled.

Reply:
I think this is reasonable.

Maybe we should consider disabling jitter if error injection is on to make the usage of this deterministic by default. It seems like an easy piece to forget, and a test author may not think they want deterministic error injection until after it fails and they go on debugging.

Reply:
Generally, I don't like it when setting one config results in sneakily setting other stuff. We can achieve the same with some ducktape utils. I don't feel too strongly about this, so if you're convinced I'm happy to do it.